driver/razerkbd_driver.c

```diff
@@ -1719,67 +1719,61 @@ static int razer_raw_event(struct hid_device *hdev, struct hid_report *report, u
     struct razer_kbd_device *asc = hid_get_drvdata(hdev);
     struct usb_device *usb_dev = interface_to_usbdev(intf);
 
-    // No translations needed on the Pro...
-    if (is_blade_laptop(usb_dev)) {
-        return 0;
-    }
-
-    // ...and on these keyboards
-    switch (usb_dev->descriptor.idProduct) {
-    case USB_DEVICE_ID_RAZER_HUNTSMAN_ELITE:
+    // No translations needed on these situations
+    if (is_blade_laptop(usb_dev) ||
+          usb_dev->descriptor.idProduct == USB_DEVICE_ID_RAZER_HUNTSMAN_ELITE ||
+          intf->cur_altsetting->desc.bInterfaceProtocol == USB_INTERFACE_PROTOCOL_KEYBOARD ||
+          (usb_dev->descriptor.idProduct != USB_DEVICE_ID_RAZER_BLACKWIDOW_2019 && size != 16) ||
+          (usb_dev->descriptor.idProduct == USB_DEVICE_ID_RAZER_BLACKWIDOW_2019 && size != 22) ||
+          data[0] != 0x04) {
```

Comment:
This is pretty unreadable :(

Can you make it so it's also "extendable" to other devices that might need the same change in the future?

Reply:
IMHO it improves the readability from previous version.
A single point of "pre-check" before proceed.
I agree on that the check it's somehow crazy... Why 16 vs 22? or What's 0x04 core means?

I don't also know what all variables mean or why HUNTSMAN_ELITE is discarded. And, more important, I can't predict next needed changes on this function.

Sorry :-)

Reply:
Maybe make an `if (usb_dev->descriptor.idProduct == USB_DEVICE_ID_RAZER_BLACKWIDOW_2019)` and in there you make the cases for the blackwidow and in the else you make a `if (!(intf->cur_altsetting->desc.bInterfaceProtocol == USB_INTERFACE_PROTOCOL_KEYBOARD && size == 16 && data[0] == 0x04))` maybe?

Reply:
i think this function may take this form (in short):
```

static int razer_raw_event(...)
{
  //1.- Preliminary checks. Return 0 if checks doesnt pass
  if (HUNTSMAN_ELITE) return 0;
  if (BW_2019 && size!=22) return 0;
  if (intf->cur_altsetting->desc.bInterfaceProtocol == USB_INTERFACE_PROTOCOL_KEYBOARD) return 0;
  if (data[0] != 0x04) return 0;
  ...

  //2.- Staff to do
  // Convert 04... to 0100...
  int index = size-1;
  ...

  return 1;
}


```
Two separate parts. No if wrapping the second part. That's my idea.
There are plenty options on how to implement first part.

I can write the first part with something like:
```
...
if (HUNTSMAN_ELITE) return 0;
if(blackwidow_2019){
  expected_size=22;
}else{
  expected_size=16;
}
if (size!=expected_size) return 0;
if (intf->cur_altsetting->desc.bInterfaceProtocol == USB_INTERFACE_PROTOCOL_KEYBOARD) return 0;
if (data[0] != 0x04) return 0;
...

```

Reply:
I'd make it
```c
if (usb_dev->descriptor.idProduct == USB_DEVICE_ID_RAZER_BLACKWIDOW_2019) {
    if (size != 22) {
        return 0;
    }
} else if (!(intf->cur_altsetting->desc.bInterfaceProtocol == USB_INTERFACE_PROTOCOL_KEYBOARD && size == 16 && data[0] == 0x04)) {
    return 0;
}

// rest of the code
```

Reply:
Shouldn't data[0] be checked in case of BLACKWIDOW2019 keyboards?

> I'd make it
> 
> ```c
> if (usb_dev->descriptor.idProduct == USB_DEVICE_ID_RAZER_BLACKWIDOW_2019) {
>     if (size != 22) {
>         return 0;
>     }
> } else if (!(intf->cur_altsetting->desc.bInterfaceProtocol == USB_INTERFACE_PROTOCOL_KEYBOARD && size == 16 && data[0] == 0x04)) {
>     return 0;
> }
> 
> // rest of the code
> ```

Reply:
I'm not familiar with the code here so I can't really say what `data[0]` is here and what should be done with it..